streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java

```diff
@@ -0,0 +1,211 @@
+
+        return repartitionTopicsMetadata;
+    }
+
+    private void ensureCopartitioning(final Collection<Set<String>> copartitionGroups,
+                                      final Map<String, InternalTopicConfig> repartitionTopicMetadata,
+                                      final Cluster clusterMetadata) {
+        for (final Set<String> copartitionGroup : copartitionGroups) {
+            copartitionedTopicsEnforcer.enforce(copartitionGroup, repartitionTopicMetadata, clusterMetadata);
+        }
+    }
+
+    private void checkIfExternalSourceTopicsExist(final TopicsInfo topicsInfo,
+                                                  final Cluster clusterMetadata) {
+        final Set<String> externalSourceTopics = new HashSet<>(topicsInfo.sourceTopics);
+        externalSourceTopics.removeAll(topicsInfo.repartitionSourceTopics.keySet());
+        externalSourceTopics.removeAll(clusterMetadata.topics());
+        if (!externalSourceTopics.isEmpty()) {
+            log.error("The following source topics are missing/unknown: {}. Please make sure all source topics " +
+                    "have been pre-created before starting the Streams application. ",
+                externalSourceTopics);
+            throw new MissingSourceTopicException("Missing source topics.");
+        }
+    }
+
+    /**
+     * Computes the number of partitions and sets it for each repartition topic in repartitionTopicMetadata
+     */
+    private void setRepartitionSourceTopicPartitionCount(final Map<String, InternalTopicConfig> repartitionTopicMetadata,
+                                                         final Map<Integer, TopicsInfo> topicGroups,
+                                                         final Cluster clusterMetadata) {
+        boolean partitionCountNeeded;
+        do {
+            partitionCountNeeded = false;
+            boolean progressMadeThisIteration = false;  // avoid infinitely looping without making any progress on unknown repartitions
+
+            for (final TopicsInfo topicsInfo : topicGroups.values()) {
+                for (final String repartitionSourceTopic : topicsInfo.repartitionSourceTopics.keySet()) {
+                    final Optional<Integer> repartitionSourceTopicPartitionCount =
+                        repartitionTopicMetadata.get(repartitionSourceTopic).numberOfPartitions();
+
+                    if (!repartitionSourceTopicPartitionCount.isPresent()) {
+                        final Integer numPartitions = computePartitionCount(
+                            repartitionTopicMetadata,
+                            topicGroups,
+                            clusterMetadata,
+                            repartitionSourceTopic
+                        );
+
+                        if (numPartitions == null) {
+                            partitionCountNeeded = true;
+                            log.trace("Unable to determine number of partitions for {}, another iteration is needed",
```

Comment:
Maybe I'm missing something, but it looks like we want to go through the `do` loop again if we have null partitions, but in this case wouldn't we hit the if statement below and throw an exception, since at this point `partitionCountNeeded == True` and `progressMadeThisIteration == False` ? Or does that just log an exception and continue through the `do` loop?

Reply:
@lct45 sort of; the point of `progressMadeThisIteration` is to break out of the loop entirely if we detect that we're infinitely looping, as was the case in a recent-ish bug. We want to continue looping only as long as we're actually gaining new information/filling in the partition count of some topic group (subtopology). 

So yes, if `progressMadeThisIteration = false` after the full loop, but we still don't have all the partition counts (`partitionCountNeeded = true`), then oh well. Something went wrong 🙂  We should throw